### PR TITLE
First commit to ft-specs

### PR DESCRIPTION
**Pull request by kbjude**
- Tests for the Tic-tac-game on the cloned repo because my partner did it solo
- Had pushed some code to the develop branch by accident

### CODE REVIEW

**Comment by shshamim63** [comment]
### Status :traffic_light: Changes Required
___
Hi @kbjude :wave: ,
well, the project contains a well descriptive readme good job :clap:  But have found some errors that are blocking me to approve it now.
___
### :spades: Mandatory Changes :spades: 
___
- How to execute the test is not in the readme. While I tried to execute the tests using `rspec spec/game_spec.rb -fdoc` it threw an error.
image
- Ther are some methods in the game class that you left without testing. <br>
   - map
   - winner?
- You can add some more scenarios to make sure that your methods are capable to adopt any scenarios(Better say alternative scenarios). For example,
https://github.com/kbjude/tic-tac-toe/blob/767922d7ff462098feef358d0ef971d680ce9133/spec/game_spec.rb#L9-L16
when `new_game.ended?` should not be true and `new_game.moves` should be greater than 0
___
##### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Comment by kbjude** [comment]
Thank you Shakhawat Hossain, this is very helpful feedback. There is where they tell us to test for public methods and the two mentioned methods above(Map and winner) are private should i go ahead and write the specs. Thank you

**Comment by shshamim63** [comment]
@kbjude the game iterations actually depend on mainly the winner? method. Don't you think that? So I  think it should not be private. And leaving an important method without testing does not seem quite right to me. What do you think?

File: spec/game_spec.rb [hunk 1]
```diff
@@ -48,24 +51,72 @@
...
+    end
+
+    context 'When the game has eneded' do
+      it 'it returns the game result' do
+        new_game.move(new_player, 7)
+        new_game.move(new_player, 3)
+        new_game.move(new_player, 5)
+        new_game.move(new_player, 1)
+        new_game.move(new_player, 9)
+        expect(new_game.moves).to eq(5)
```
Comment (kobiyoyo) [inline spec/game_spec.rb]: You should have one `it` block for each expected behaviour.

File: spec/game_spec.rb [hunk 2]
```diff
@@ -5,20 +5,23 @@
...
     end
   end
 
   describe '#player_move' do
-    context 'When a player wants to make a move' do
-      it 'When a new player makes a move' do
+    context 'When Playing' do
+      it 'a new player makes a move' do
         expect(new_game.move(new_player, 3)).to eql(nil)
+        expect(new_game.move(new_player, 1)).should_not eq(0)
```
Comment (kobiyoyo) [inline spec/game_spec.rb:24]: You should have one it block for each expected behaviour.

File: spec/game_spec.rb [hunk 3]
```diff
@@ -48,24 +51,72 @@
         new_game.move(new_player, 1)
         new_game.move(new_player, 9)
         expect(new_game.ended?).to eql(true)
+        expect(new_game.ended?).should_not eq(new_game.move(new_player, 0))
```
Comment (kobiyoyo) [inline spec/game_spec.rb]: You should have one it block for each expected behaviour.

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @kbjude     
#### Great Job so far, but there are a few more things required. 
- [ ] Lets write test for when there is a draw between both players

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by kbjude** [comment]
All feedback implemented

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @kbjude     
#### Great Job so far, but there is one more thing  required. 
- [ ] Lets write test for when there is a draw between both players

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by kbjude** [comment]
Player's tie added

**Comment by Anna-Myzukina** [comment]
# Hi @kbjude ! :clap: 
### Good job   :eyes: 



* [ ] All the public methods of the classes implemented in project OOP should be tested => from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/)
screenshot 

You could split your test between different files following the file that you're testing, so you'll have something like: 
- game_spec.rb, 
- player_spec.rb, 
- userinterface_spec.rb, 
that way it is easy to find the test cases for each class/module and you can implement more granular unit tests on methods for each file. 


### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*
*Happy coding * <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

**Comment by kbjude** [comment]
- Hi @Anna-Muzykina thank you for the feedback, could please mention the public methods that are lacking specs, i feel i have exhausted them.
- About the granular approach, is it a must have, i feel i will get lost in the many files, if it is, please re-echo that so that i can implement that but all specs for the listed classes above are in the game_spec.rb

**Review by majovanilla — APPROVED** [review]
Hi team!

## Good job on this project! 👍

I will approve your pull request with just a few suggestions to keep in mind for future projects:

- There's absolutely no need to test your private methods. As long as the public method that it's calling them is working, you don't have to touch the private one, no matter how "important" it is.

- There is no need to convert private methods into public! The main reason to have private methods is to keep your code secure and unreachable if there is no need to access it from outside your class. 

- You should always try to keep your tests (and any other code) in different files when they are too long. Like Anna said, it's better to separate the tests. Perhaps right now it might seem unnecessary to you, but in the future, you will have larger projects and trust me, it will be easier that way.

- Imagine your test is a sentence: it should be read easily. In some tests, you have things like this: `    it 'when the player are tieing' do` That is not very readable! You could improve the line using something like this: `    it 'shows when the game is a tie' do`

- To contact TSE's is always better to reach out on Slack. We usually never see replies here because we get A LOT of emails from github from all the projects we've been reviewing (I don't check all of the emails) So to be sure, send us a message!


If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

**Pull request merged**